Sources/SwiftCompilerPluginMessageHandling/JSON/JSONDecoding.swift

```diff
@@ -0,0 +1,1286 @@
+        if advance(if: ",") {
+          skipWhitespace()
+          continue
+        }
+        break
+      }
+      try expect("]")
+    }
+    map.closeCollection(handle: handle)
+  }
+
+  mutating func scanValue() throws {
+    skipWhitespace()
+    let start = ptr
+    switch try advance() {
+    case UInt8(ascii: "n"):
+      try scanNull()
+    case UInt8(ascii: "t"):
+      try scanTrue()
+    case UInt8(ascii: "f"):
+      try scanFalse()
+    case UInt8(ascii: "\""):
+      try scanString(start: start)
+    case UInt8(ascii: "-"), UInt8(ascii: "0")...UInt8(ascii: "9"):
+      try scanNumber(start: start)
+    case UInt8(ascii: "{"):
+      try scanObject()
+    case UInt8(ascii: "["):
+      try scanArray()
+    case let chr:
+      throw JSONError.unexpectedCharacter(chr, context: "value start")
+    }
+    skipWhitespace()
+  }
+
+  static func scan(buffer: UnsafeBufferPointer<UInt8>) throws -> JSONMap {
+    var scanner = JSONScanner(buffer: buffer)
+    try scanner.scanValue()
+    if scanner.hasData {
+      throw JSONError.unexpectedCharacter(scanner.ptr.pointee, context: "after top-level value")
+    }
+
+    return scanner.map.finalize()
+  }
+}
+
+/// Represents a single value in a JSONMap.
+private struct JSONMapValue {
+  typealias Pointer = UnsafePointer<Int>
+  /// Pointer to an element descriptor in the map.
+  let data: Pointer
+
+  @inline(__always)
+  var endPtr: Pointer {
+    data.advanced(by: mapSize)
+  }
+
+  /// The size of this value data in the map.
+  @inline(__always)
+  var mapSize: Int {
+    switch JSONMap.Descriptor(rawValue: data[0]) {
+    case .nullKeyword, .trueKeyword, .falseKeyword:
+      return 1
+    case .number, .asciiSimpleString, .simpleString, .string:
+      return 3
+    case .array, .object:
+      return data[1]
+    case nil:
+      fatalError("invalid value descriptor")
+    }
+  }
+
+  @inline(__always)
+  func `is`(_ kind: JSONMap.Descriptor) -> Bool {
+    return data[0] == kind.rawValue
+  }
+}
+
+// MARK: Keyword primitives
+extension JSONMapValue {
+  @inline(__always)
+  var isNull: Bool {
+    return self.is(.nullKeyword)
+  }
+
+  @inline(__always)
+  func asBool() -> Bool? {
+    if self.is(.trueKeyword) {
+      return true
+    }
+    if self.is(.falseKeyword) {
+      return false
+    }
+    return nil
+  }
+}
+
+// MARK: Scalar values
+private enum _JSONStringParser {
+  /// Decode a non-escaped string value from the buffer.
+  @inline(__always)
+  static func decodeSimpleString(source: UnsafeBufferPointer<UInt8>) -> String {
+    if source.count <= 0 {
+      return ""
+    }
+    return _makeString(unsafeUninitializedCapacity: source.count) { buffer in
+      buffer.initialize(fromContentsOf: source)
+    }
+  }
+
+  /// Decode a string value that includes escape sequences.
+  static func decodeStringWithEscapes(source: UnsafeBufferPointer<UInt8>) -> String? {
+    // JSON string with escape sequences must be 2 bytes or longer.
+    assert(source.count > 0)
```

Comment:
Yes, `> 0` is enough for assertion because I just want to ensure `baseAddress` is not null.
This is like "since escaped sequence is 2 bytes or longer, this string must not be empty"

Reply:
Ah, now I see. The comment is the precondition and the assertion is what follows from it.